src/mlpack/methods/kmeans/naive_kmeans_impl.hpp

```diff
@@ -38,28 +40,52 @@ double NaiveKMeans<MetricType, MatType>::Iterate(const arma::mat& centroids,
   counts.zeros(centroids.n_cols);
 
   // Find the closest centroid to each point and update the new centroids.
-  for (size_t i = 0; i < dataset.n_cols; i++)
+  // Computed in parallel over the complete dataset
+  #pragma omp parallel
   {
-    // Find the closest centroid to this point.
-    double minDistance = std::numeric_limits<double>::infinity();
-    size_t closestCluster = centroids.n_cols; // Invalid value.
+    // The current state of the K-means is private for each thread
+    arma::mat localCentroids(centroids.n_rows, centroids.n_cols,
+                             arma::fill::zeros);
+    arma::Col<size_t> localCounts(centroids.n_cols, arma::fill::zeros);
 
-    for (size_t j = 0; j < centroids.n_cols; j++)
+#ifdef _WIN32
+  // Tiny workaround: Visual Studio only implements OpenMP 2.0, which doesn't
+  // support unsigned loop variables. If we're building for Visual Studio, use
+  // the intmax_t type instead.
+  #pragma omp for
+    for (intmax_t i = 0; i < (intmax_t) dataset.n_cols; ++i)
+#else
+  #pragma omp for
+    for (size_t i = 0; i < dataset.n_cols; ++i)
+#endif
     {
-      const double distance = metric.Evaluate(dataset.col(i), centroids.col(j));
+      // Find the closest centroid to this point.
+      double minDistance = std::numeric_limits<double>::infinity();
+      size_t closestCluster = centroids.n_cols; // Invalid value.
 
-      if (distance < minDistance)
+      for (size_t j = 0; j < centroids.n_cols; j++)
       {
-        minDistance = distance;
-        closestCluster = j;
+        const double distance = metric.Evaluate(dataset.col(i),
+                                                centroids.unsafe_col(j));
+        if (distance < minDistance)
+        {
+          minDistance = distance;
+          closestCluster = j;
+        }
       }
-    }
 
-    Log::Assert(closestCluster != centroids.n_cols);
+      Log::Assert(closestCluster != centroids.n_cols);
 
-    // We now have the minimum distance centroid index.  Update that centroid.
-    newCentroids.col(closestCluster) += arma::vec(dataset.col(i));
-    counts(closestCluster)++;
+      // We now have the minimum distance centroid index.  Update that centroid.
+      localCentroids.unsafe_col(closestCluster) += dataset.col(i);
+      localCounts(closestCluster)++;
+    }
+    // Combine calculated state from each thread
+    #pragma omp critical
```

Comment:
I am not too sure about how that would make a reduction possible. Any variable inside the reduction clause is restricted to be a scalar (OpenMP does not define what scalar means, but it most probably  means numeric POD types).
Each thread calculates its share of centroids and counts, so the reduction must be able to sum up the centroids and counts which are vectors themselves. I can't think of a way to achieve this in the  absence of user-defined reductions.
I'll measure how much time is spent on waiting for the critical section to unlock.